src/MonoTorrent/MonoTorrent.Streaming/LocalStream.cs

```diff
@@ -0,0 +1,155 @@
+    /// If the stream seeks to a location which hasn't been downloaded yet, <see cref="Read(byte[], int, int)"/>
+    /// will block until the data is available. <see cref="ReadAsync(byte[], int, int, CancellationToken)"/>
+    /// will perform a non-blocking wait for the data.
+    /// </summary>
+    class LocalStream : Stream
+    {
+        long position;
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        public override long Length => File.Length;
+
+        public override long Position {
+            get => position;
+            set => Seek (value, SeekOrigin.Begin);
+        }
+
+        TorrentFile File { get; }
+
+        TorrentManager Manager { get; }
+
+        StreamingPiecePicker Picker { get; }
+
+        FileStream Stream { get; set; }
+
+        public LocalStream (TorrentManager manager, TorrentFile file, StreamingPiecePicker picker)
+        {
+            Manager = manager;
+            File = file;
+            Picker = picker;
+        }
+
+        protected override void Dispose (bool disposing)
+        {
+            base.Dispose (disposing);
+            if (disposing) {
+                Stream?.Dispose ();
+            }
+        }
+
+        public override void Flush ()
+            => throw new NotSupportedException ();
+
+        public override int Read (byte[] buffer, int offset, int count)
+            => ReadAsync (buffer, offset, count, CancellationToken.None).GetAwaiter ().GetResult ();
+
+        public override async Task<int> ReadAsync (byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // The torrent is treated as one big block of data, so this is the offset at which the current file's data starts at.
+            var torrentFileStartOffset = (long)File.StartPieceIndex * (long)Manager.Torrent.PieceLength + File.StartPieceOffset;
+
+            // Take our current position into account when calculating the start/end pieces of the data we're reading.
+            var startPiece = (int) (torrentFileStartOffset + Position) / Manager.Torrent.PieceLength;
+            var endPiece = (int) (torrentFileStartOffset + Position + count) / Manager.Torrent.PieceLength;
+            while (Manager.State != TorrentState.Stopped && Manager.State != TorrentState.Error) {
+                bool allAvailable = true;
+                for (int i = startPiece; i <= endPiece && allAvailable; i++)
+                    allAvailable &= Manager.Bitfield[i];
+
+                if (allAvailable)
+                    break;
+
+                await Task.Delay (500, cancellationToken);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested ();
+            if (Stream == null) {
+                Stream = new FileStream (File.FullPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
```

Comment:
I guess you can use `DiskManager.Read` when you'll refactor this code

Reply:
Hah, that's a very good point! I should definitely fix that :) I suspect that would simplify things further too :)

This is something I did while prototyping in a standalone application and I never even thought to move to DiskManager! Oops :)

Great catch!